packages/ciphernode/net/src/network_peer.rs

```diff
@@ -29,135 +51,254 @@ pub struct NodeBehaviour {
+                        .gossipsub
+                        .add_explicit_peer(&peer_id);
+                    info!("Added peer to gossipsub {}", remote_addr);
+                    mgr.do_send(NetworkPeerEvent::ConnectionEstablished { connection_id });
+                }
+
+                SwarmEvent::OutgoingConnectionError {
+                    peer_id,
+                    error,
+                    connection_id,
+                } => {
+                    info!("Failed to dial {peer_id:?}: {error}");
+                    if let Some(dialer) = &mut self.dialer {
+                        dialer.do_send(NetworkPeerEvent::OutgoingConnectionError {
+                            connection_id,
+                            error: Arc::new(error),
+                        });
+                    }
+                }
+
+                SwarmEvent::IncomingConnectionError { error, .. } => {
+                    warn!("{:#}", anyhow::Error::from(error))
+                }
+
+                SwarmEvent::Behaviour(NodeBehaviourEvent::Kademlia(e)) => {
+                    debug!("Kademlia event: {:?}", e);
+                }
+
+                SwarmEvent::Behaviour(NodeBehaviourEvent::Mdns(mdns::Event::Discovered(list))) => {
+                    for (peer_id, _multiaddr) in list {
+                        trace!("mDNS discovered a new peer: {peer_id}");
+                        self.swarm
+                            .behaviour_mut()
+                            .gossipsub
+                            .add_explicit_peer(&peer_id);
+                    }
+                }
+
+                SwarmEvent::Behaviour(NodeBehaviourEvent::Mdns(mdns::Event::Expired(list))) => {
+                    for (peer_id, _multiaddr) in list {
+                        trace!("mDNS discover peer has expired: {peer_id}");
+                        self.swarm
+                            .behaviour_mut()
+                            .gossipsub
+                            .remove_explicit_peer(&peer_id);
+                    }
+                }
+
+                SwarmEvent::Behaviour(NodeBehaviourEvent::Gossipsub(
+                    gossipsub::Event::Message {
+                        propagation_source: peer_id,
+                        message_id: id,
+                        message,
+                    },
+                )) => {
+                    trace!("Got message with id: {id} from peer: {peer_id}",);
+                    mgr.do_send(NetworkPeerEvent::GossipData(message.data));
+                }
+
+                SwarmEvent::NewListenAddr { address, .. } => {
+                    info!("Local node is listening on {address}");
+                }
+                _ => {}
+            };
+        }
+        Ok(())
     }
+}
+
+impl Actor for NetworkPeer {
+    type Context = Context<Self>;
+
+    fn started(&mut self, ctx: &mut Self::Context) {
+        // Setup stream handling for swarm events
+        ctx.add_stream(Box::pin(futures::stream::unfold(
```

Comment:
cool!

Reply:
yeah but sadly not possible, have to resort to polling.